File: CUDA Implementation MacOSX/CUDA_Tomography.m
clear
clc

% Load Image and Give Pixels (x,y) Coordinates
orig_object = im2double(rgb2gray(imread('../images/cholangioca.jpg')));
[Ny, Nx] = size(orig_object); 
dx = 0.1; dy = 0.1;
x = (-(Nx-1)/2:(Nx-1)/2) * dx;
y = (-(Ny-1)/2:(Ny-1)/2) * dy;

% Determining Sensor Array Element Spacing and Extent
dr = 1/(1/dx + 1/dy);
rmax = sqrt(max(abs(x))^2 + max(abs(y))^2);
rmax = dr*ceil(rmax/dr); 
Nr = round(2*rmax/dr + 1);

% Projection Angles for Computed Tomography
numAngles = 360; totalCoverage = 180;
theta = linspace(0,totalCoverage-totalCoverage/numAngles,numAngles);

% Save Important Values to .txt files before running CUDA
save_orig_object = orig_object'; % Just make row dimension the fast dimension
fid_img = fopen('img.txt','w');
fprintf(fid_img,'%f\n', save_orig_object(:));
fid_theta = fopen('theta.txt','w');
fprintf(fid_theta,'%f\n', theta(:));
fclose(fid_img); fclose(fid_theta);

% Compile then run the CUDA program that creates the sinogram
system(['./sinogram.out ', num2str(numAngles), ' theta.txt ', ...
	num2str(-rmax), ' ', num2str(rmax), ' ', num2str(Nr), ' sg.txt ', ...
	num2str(min(x)), ' ', num2str(max(x)), ' ', num2str(Nx), ' ', ...
	num2str(min(y)), ' ', num2str(max(y)), ' ', num2str(Ny), ' img.txt']);

% Compile then run the CUDA program that performs the filtered backprojection
system(['./filtBackproj.out ', num2str(numAngles), ' theta.txt ', ...
	num2str(-rmax), ' ', num2str(rmax), ' ', num2str(Nr), ' sg.txt ', ...
	num2str(min(x)), ' ', num2str(max(x)), ' ', num2str(Nx), ' ', ...
	num2str(min(y)), ' ', num2str(max(y)), ' ', num2str(Ny), ' recon.txt']);

% Load the reconstructed image
recon_img = textread('recon.txt');
recon_img = reshape(recon_img, Nx, Ny)';

% Delete all files that this process created
system('rm img.txt');
system('rm recon.txt');
system('rm sg.txt');
system('rm theta.txt');

% Show the reconstructed images
imagesc(x, y, recon_img);
colormap gray;

File: CUDA Implementation MacOSX/CUDA_Tomography.py
import numpy as np
import matplotlib.pyplot as plt
from scipy import misc
import os, pdb

# Load Image and Give Pixels (x,y) Coordinates
orig_object = misc.imread('../images/cholangioca.jpg', flatten = True);
Ny, Nx = orig_object.shape; 
dx = 0.1; dy = 0.1;
x = (np.arange(Nx) - (Nx-1)/2) * dx;
y = (np.arange(Ny) - (Ny-1)/2) * dy;

# Determining Sensor Array Element Spacing and Extent
dr = 1/(1/dx+ 1/dy);
rmax = np.sqrt(np.max(np.abs(x))**2 + np.max(np.abs(y))**2);
rmax = dr*np.ceil(rmax/dr); 
Nr = int(2*rmax/dr + 1);

# Projection Angles for Computed Tomography
numAngles = 360;
theta = np.linspace(0,180,numAngles,endpoint=False);

# Save Important Values to .txt files before running CUDA
np.savetxt('img.txt', orig_object.flatten());
np.savetxt('theta.txt', theta);

pdb.set_trace();

# Compile then run the CUDA program that creates the sinogram
os.system('nvcc sinogram.cu -o sinogram.out');
os.system('./sinogram.out '+str(numAngles)+' theta.txt '+ \
	str(-rmax)+' '+str(rmax)+' '+str(Nr)+' sg.txt '+ \
	str(np.min(x))+' '+str(np.max(x))+' '+str(Nx)+' '+ \
	str(np.min(y))+' '+str(np.max(y))+' '+str(Ny)+' img.txt');

# Compile then run the CUDA program that performs the filtered backprojection
os.system('nvcc filtBackproj.cu -o filtBackproj.out -I/usr/local/cuda/include -L/usr/local/cuda/lib -lcufft');
os.system('./filtBackproj.out '+str(numAngles)+' theta.txt '+ \
	str(-rmax)+' '+str(rmax)+' '+str(Nr)+' sg.txt '+ \
	str(np.min(x))+' '+str(np.max(x))+' '+str(Nx)+' '+ \
	str(np.min(y))+' '+str(np.max(y))+' '+str(Ny)+' recon.txt');

# Load the reconstructed image
recon_img = np.loadtxt('recon.txt');
recon_img = np.reshape(recon_img, (Ny,Nx))

# Delete all files that this process created
os.system('rm img.txt');
os.system('rm recon.txt');
os.system('rm sg.txt');
os.system('rm theta.txt');

# Show the reconstructed images
plt.imshow(recon_img, cmap='gray');
plt.show();
